src/DAG/dag.c

```diff
@@ -573,10 +536,98 @@ void RedisAI_BatchedDagRunSessionStep(RedisAI_RunInfo **batched_rinfo, const cha
+        }
+    }
+    ret = REDISMODULE_OK;
+clean_up:
+    RedisModule_CloseKey(key);
+    RedisAI_ReplicateTensorSet(ctx, tensor_keyname, tensor);
+    return ret;
+}
+
+static void PersistTensors(RedisModuleCtx *ctx, RedisAI_RunInfo *rinfo) {
+    AI_dictIterator *persist_iter = AI_dictGetSafeIterator(rinfo->dagTensorsPersistedContext);
+    AI_dictEntry *persist_entry = AI_dictNext(persist_iter);
+    while (persist_entry) {
+        const char *persist_key_name = AI_dictGetKey(persist_entry);
+
+        AI_dictEntry *tensor_entry = AI_dictFind(rinfo->dagTensorsContext, persist_key_name);
+
+        if (tensor_entry) {
+            RAI_Tensor *tensor = AI_dictGetVal(tensor_entry);
+
+            if (tensor == NULL) {
+                persist_entry = AI_dictNext(persist_iter);
+                continue;
+            }
+            bool mangled = true;
+            if (StoreTensorInKeySpace(ctx, tensor, persist_key_name, mangled) == REDISMODULE_ERR)
+                rinfo->dagReplyLength++;
+
+        } else {
+            RedisModule_ReplyWithError(ctx,
+                                       "ERR specified persistent key that was not used in DAG");
+            rinfo->dagReplyLength++;
+
+            RedisModule_Log(ctx, "warning",
+                            "on DAGRUN's PERSIST pecified persistent key (%s) that "
+                            "was not used on DAG. Logging all local context keys",
+                            persist_key_name);
+            AI_dictIterator *local_iter = AI_dictGetSafeIterator(rinfo->dagTensorsContext);
+            AI_dictEntry *local_entry = AI_dictNext(local_iter);
+            while (local_entry) {
+                const char *localcontext_key_name = AI_dictGetKey(local_entry);
+                RedisModule_Log(ctx, "warning", "DAG's local context key (%s)",
+                                localcontext_key_name);
+                local_entry = AI_dictNext(local_iter);
+            }
+            AI_dictReleaseIterator(local_iter);
+
+            for (size_t opN = 0; opN < array_len(rinfo->dagOps); opN++) {
+                RedisModule_Log(ctx, "warning", "DAG's op n#  %zu - cmdType %d ( argc %d )", opN,
+                                rinfo->dagOps[opN]->commandType, rinfo->dagOps[opN]->argc);
+            }
+        }
+        persist_entry = AI_dictNext(persist_iter);
+    }
+    AI_dictReleaseIterator(persist_iter);
+}
+
+static void ModelSingleOp_PersistTensors(RedisModuleCtx *ctx, RAI_DagOp *op) {
+    const size_t noutputs = RAI_ModelRunCtxNumOutputs(op->mctx);
+    for (size_t outputNumber = 0; outputNumber < noutputs; outputNumber++) {
+        RAI_Tensor *tensor = RAI_ModelRunCtxOutputTensor(op->mctx, outputNumber);
+        const char *key_string = RedisModule_StringPtrLen(op->outkeys[outputNumber], NULL);
+        tensor = tensor ? RAI_TensorGetShallowCopy(tensor) : NULL;
+        bool mangled = false;
+        if (tensor)
+            StoreTensorInKeySpace(ctx, tensor, key_string, mangled);
```

Comment:
In which situations we will have `NULL` here and why we continue without errors?